### DISCUSSION TITLE
Vue: Improve code snippets

### DISCUSSION THREAD

**Discussion by larsrickert** [storybookjs/storybook#26691] [Ideas]
### Is your feature request related to a problem? Please describe.

We are using Storybook to document our Vue component library. Its very useful the the code snippets are automatically generated so the user can copy it to his project.

However, we would like to have more modern code snippets that better follow common Vue best practices / are closer to the code that the user would actually use in his project. This includes:

- use `@` shorthand for events instead of `v-on:`
- use shorthand for `true` boolean props
- move array/object values to `<script>`

### Describe the solution you'd like

I can implement the change but I want to clarify/discuss first if this feature request is also a desired change on your side.

Example:

**Current snippet**:
```vue
<template>
  <MyComponent
    :value="{ foo: 42, bar: 'baz' }"
    :disabled="true"
    v-on:some-event="() => ({})"
  />
</template>
```

**My suggestion**:
```vue
<script lang="ts" setup>
const value = { foo: 42, bar: "baz" };
</script>

<template>
  <MyComponent :value="value" disabled @some-event="() => ({})" />
</template>
```

Also I'd like to have a `Code` tab added when viewing a single story. Currently the source code is only available on the "Docs" page. Example:
<img width="659" alt="Screenshot 2024-03-29 at 16 32 06" src="https://github.com/storybookjs/storybook/assets/67898185/919c89d9-1c63-4944-8d65-3fb7ae3ad937">

### Are you able to assist to bring the feature to reality?

yes, I can

### Additional context

The feature could be implemented as opt-in framework option for Storybook 8 and can be changed to be the default in Storybook 9 so we don't introduce possible breaking changes for projects using custom code snippet transformers.

**Comment by larsrickert** [comment]
@kasperpeulen @chakAs3 Would love your opinion on this :)

**Reply by kasperpeulen** [reply]
This looks great to me.
Tagging @ruphin as he also had some ideas here I belief.

**Reply by ruphin** [reply]
Looks good, the more idiomatic Vue code it produces the better.

In general, any improvement to the example "code" for Vue is quite welcome, as it tends to not be very copy-paste friendly at the moment.

**Reply by chakAs3** [reply]
It is good initiative it was in my backlog, thanks again @larsrickert , and for copy-paste friendly as @ruphin i have tried my best however i was waiting for users feedback but at least i claim that when you copy past the code it works still for idiomatic code i agree we should discuss this for example : 
when you write a story using  v-bind your story code 
```ts
export const Overview = {
  args: { size:'medium' , color:'red' },
  render: (args: any) => ({
    setup: () => {
      return { args }
    },
    template: ` <v-btn v-bind="args">Button</v-btn>`
  })
}
```
so you have to know there is difference between Story source which is normally shows on the tab and the generated template that you can copy past in your Vue app in this case the previous code is the Story source the following code is to options both of them should work 

Option 1

```vue 
<template>
    <v-btn v-bind="{ size:'medium' , color:'red' }">Button</v-btn>
</template>
```

or Option 2

```vue
<template>
   <v-btn  size="medium" color="red" >Button</v-btn>
</template>
```

Option 1:  translates what you have really use which is _**v-bind**_ 
Option 2:  more idiomatic Vue code that you usually find or use 

There is even one more option is using dynamic _ref_ so you will have a script tag + template tag: this option is available if you pass an additional tag to parameters  
```vue
<script>
   const size = ref('medium')
   const color = ref('red')
</script>
<template>
   <v-btn  :size="size" :color="color" >Button</v-btn>
</template>

**Reply by larsrickert** [reply]
Thanks for the explanation :) I think a combination between Option 2 and 3 is what we should aim for.
So use regular props like `size="medium"` for primitive types but move non-primitive types like objects, arrays and functions to the `<script>` section.

**Comment by chakAs3** [comment]
> Also I'd like to have a `Code` tab added when viewing a single story. Currently the source code is only available on the "Docs" page.

Point 1: There's this addon called Storysource, but it's aimed at displaying the source code of the story, not the generated code.

Point 2: Bringing in this feature wouldn't just be a win for Vue users; it'd be a big deal for folks using other frameworks too. I reckon a bunch of users would really appreciate having this in their toolkit.

Point 3: This feature was actually what drove me to start contributing in the first place. It's what got me hooked, and now I'm the one maintaining it. happy to have @larsrickert joining me on same path.

**Comment by larsrickert** [comment]
I will close this because #27194 and #29253 are already implemented and merged 🎉

**Reply by yashaswi1004** [reply]
hey can you tell the detailed steps of how the complete source code is generated and displayed?
because when using vite framework and storysource addon we are not able to generate the source code.

**Reply by larsrickert** [reply]
Hey, the source code is generated as described below. Note that currently the whole source code generation is only based in the story args and the component docgen info (slot names etc.). Custom render functions are not supported.

1. Component name is extracted from the given Vue component
2. Props are generated based on the story args and depending on the arg type. When objects, arrays or v-models are used as value, they will be generated inside the `<script lang="ts" setup>` section instead of inlining them in the `<template>`
3. If slots are used/passed as arg, e.g. via `{ slotName: "Slot content..." }` or `{ slotName: h(MyVueComponent, { propName: 'Foo bar' }) }`, the source code for the slots are generated:
    - if slot bindings exists, they are generated accordingly, e.g. when passing arg "{ slotName: ({ data } => `Example ${data}`) }", the output will be `<template #slotName="{ data }">Example {{ data }} </template>` as expected
    - the slot content is generated recursively by repeating steps 1-3

Example Story:

```ts
import type { Meta, StoryObj } from "@storybook/vue3";
import MyComponent from "./MyComponent.vue";

const meta: Meta<typeof MyComponent> = {
  component: MyComponent,
};

export default meta;
type Story = StoryObj<typeof MyComponent>;

export const MyStory = {
  args: {
    propA: "Hello World",
    propB: 42,
    propC: [1, 2, 3],
    slotA: "Slot content...",
    slotB: h("div", { class: "my-class" }),
    slotC: ({ data }) => [
      `Test ${data}`,
      h(MyVueComponent, { someProp: "foo" }, "Component slot content"),
    ],
  },
} satisfies Story;
```

Generated source code:

```vue
<script lang="ts" setup>
import { ref } from "vue";

const propC = [1, 2, 3];
</script>

<template>
  <MyComponent propA="Hello World" :propB="42" :propC="propC">
    <template #slotA>Slot content...</template>

    <template #slotB>
      <div class="my-class">Div content</div>
    </template>

    <template #slotC="{ data }">
      Test {{ data }}
      <SomeOtherComponent someProp="foo">
        Component slot content
      </SomeOtherComponent>
    </template>
  </MyComponent>
</template>
```

**Reply by yashaswi1004** [reply]
Thanks for the reply. In the mean time i read one discussion and found that we can now use addon-docs to display the source code in the code panel. And the current issue is [this](https://github.com/SchwarzIT/onyx/issues/2379).  I am using react/vite framework. Is there a way now to solve it?

**Reply by larsrickert** [reply]
@yashaswi1004 I think this has to be fixed in Storybook itself. I already took a Quick Look but it seems like function parameters are not passed to the addon options. I‘d need more time to investigate this.

**Reply by yashaswi1004** [reply]
@larsrickert thanks for the update! I'll keep an eye out for the updates
